src/processor/operator/order_by/sort_state.cpp

```diff
@@ -107,53 +108,85 @@ PayloadScanner::PayloadScanner(MergedKeyBlocks* keyBlockToScan,
-                    auto ft = payloadTables[OrderByKeyEncoder::getEncodedFTIdx(payloadInfo)];
-                    tuplesToRead[numTuplesRead + i] =
-                        ft->getTuple(blockIdx * ft->getNumTuplesPerBlock() + blockOffset);
-                    blockPtrInfo->curTuplePtr += keyBlockToScan->getNumBytesPerTuple();
-                }
-                blockPtrInfo->updateTuplePtrIfNecessary();
-                numTuplesRead += numTuplesToReadInCurBlock;
+        auto numTuplesToRead = std::min(DEFAULT_VECTOR_CAPACITY,
+            endTuplesIdxToReadInMergedKeyBlock - nextTupleIdxToReadInMergedKeyBlock);
+        auto numTuplesRead = 0u;
+        while (numTuplesRead < numTuplesToRead) {
+            auto numTuplesToReadInCurBlock = std::min(
+                numTuplesToRead - numTuplesRead, blockPtrInfo->getNumTuplesLeftInCurBlock());
+            for (auto i = 0u; i < numTuplesToReadInCurBlock; i++) {
+                auto payloadInfo = blockPtrInfo->curTuplePtr + payloadIdxOffset;
+                auto blockIdx = OrderByKeyEncoder::getEncodedFTBlockIdx(payloadInfo);
+                auto blockOffset = OrderByKeyEncoder::getEncodedFTBlockOffset(payloadInfo);
+                auto ft = payloadTables[OrderByKeyEncoder::getEncodedFTIdx(payloadInfo)];
+                tuplesToRead[numTuplesRead + i] =
+                    ft->getTuple(blockIdx * ft->getNumTuplesPerBlock() + blockOffset);
+                blockPtrInfo->curTuplePtr += keyBlockToScan->getNumBytesPerTuple();
             }
-            // TODO(Ziyi): This is a hacky way of using factorizedTable::lookup function,
-            // since the tuples in tuplesToRead may not belong to factorizedTable0. The
-            // lookup function doesn't perform a check on whether it holds all the tuples in
-            // tuplesToRead. We should optimize this lookup function in the orderByScan
-            // optimization PR.
-            payloadTables[0]->lookup(
-                vectorsToRead, colsToScan, tuplesToRead.get(), 0, numTuplesToRead);
-            nextTupleIdxToReadInMergedKeyBlock += numTuplesToRead;
-            return numTuplesRead;
+            blockPtrInfo->updateTuplePtrIfNecessary();
+            numTuplesRead += numTuplesToReadInCurBlock;
+        }
+        // TODO(Ziyi): This is a hacky way of using factorizedTable::lookup function,
+        // since the tuples in tuplesToRead may not belong to factorizedTable0. The
+        // lookup function doesn't perform a check on whether it holds all the tuples in
+        // tuplesToRead. We should optimize this lookup function in the orderByScan
+        // optimization PR.
+        payloadTables[0]->lookup(vectorsToRead, colsToScan, tuplesToRead.get(), 0, numTuplesToRead);
+        nextTupleIdxToReadInMergedKeyBlock += numTuplesToRead;
+        return numTuplesRead;
+    }
+}
+
+bool PayloadScanner::scanSingleTuple(std::vector<common::ValueVector*> vectorsToRead) const {
```

Comment:
This is a pure compile time check. Open an issue for me to fix.

Reply:
#2652